serveradmin/graphite/management/commands/cache_graphite.py

```diff
@@ -116,10 +117,14 @@ def cache_numerics(self, collection, server):
             # it is set up like this.  This process takes a long time.
             # We want the values to be immediately available to the users.
             with transaction.atomic():
-                ServerNumberAttribute.objects.update_or_create(
-                    server_id=server.object_id,
+                # Lock server for changes to avoid nonrepeatable reads in the
+                # query_committer.
+                locked_server = Server.objects.select_for_update().get(
```

Comment:
Do I understand it correctly, that this works before it would always first try and lock the server and if not wait for it? If some code would not try and first lock the server but directly work on the attributes again it would again fail, right?
Just checking if I understood it correctly right now.

Reply:
Yes exactly. It would wait for a lock (same as in the query_committer) instead of just updating the attribute possibliy causing a nonrepeatable read in the query committer.

And yes you are right it could happen again if any other code would not follow this "protocol". Hence the comment in the query committer to evaluate if we can change the transaction level to "repeatable read" so that it is robust against other code places.